File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.ArrayDeque;
import java.util.Queue;

public class Solution {

    public int solution(String begin, String target, String[] words) {
        boolean[] visited = new boolean[words.length];
        Queue<String> queue = new ArrayDeque<>();

        queue.offer(begin);

        int count = 0;

        while (!queue.isEmpty()) {
            int levelSize = queue.size();

            for (int i = 0; i < levelSize; i++) {
                String current = queue.poll();

                if (current.equals(target)) {
                    return count;
                }

                for (int j = 0; j < words.length; j++) {
                    if (!visited[j] && canConvert(current, words[j])) {
                        visited[j] = true;
                        queue.offer(words[j]);
                    }
                }
            }

            count++;
        }

        return 0;
    }

    private boolean canConvert(String first, String second) {
        int difference = 0;

        for (int i = 0; i < first.length(); i++) {
            if (first.charAt(i) != second.charAt(i)) {
                difference++;

                if (difference > 1) {
                    return false;
                }
            }
        }

        return difference == 1;
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque


def solution(begin, target, words):
    if target not in words:
        return 0
    
    q = deque([(begin, 0)])
    
    while q:
        current_word, count = q.popleft()
        
        if current_word == target:
            return count
        
        target_words = [word for word in words if len(current_word) - 1 == sum(c == w for c, w in zip(current_word, word))]
        
        for word in target_words:
            q.append((word, count + 1))
        
    return 0
